### PR TITLE
Add psalm and type all vars

### PR DESCRIPTION
**Pull request by fezfez**
All vars are now totally typed. 
We are now 100% sure that nothing will break out of or control. 

This PR introduce 

- psalm
- a new class `RegexParts` wich represent a partials regex
- a new class `RegexMatch` wich is used to get data from preg_match

i've done some benchmark and this implementation is faster

### CODE REVIEW

File: src/Parser/RegexMatch.php [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+     * @param array<array-key, string>  $matches
+     */
+    public function __construct(array $regexParts, array $matches)
+    {
+        $values = [];
+
+        foreach (array_keys($regexParts) as $i => $key) {
+            $buildKey = $i + 1;
+            if (!array_key_exists($buildKey, $matches)) {
+                throw new RuntimeException();
```
Comment (guillaume-sainthillier) [inline src/Parser/RegexMatch.php]: Could you add a message to this exception? Or even a dedicated exception class

File: src/Parser/RegexMatch.php [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+            $values[$key] = 0 != strlen($value) ? $value : null;
+        }
+
+        $this->values = $values;
+    }
+
+    public function getString(string $key): string
+    {
+        if (!array_key_exists($key, $this->values) || !is_string($this->values[$key])) {
+            throw new RuntimeException(sprintf('key "%s" is not a string', $key));
```
Comment (guillaume-sainthillier) [inline src/Parser/RegexMatch.php:52]: Could you split your test to give a custom message for each case? if array_key_exists returns false message would be key "%s" is not a string, which is not true.

Moreover I would change 'key "%s" is not a string' by 'The value of key "%s" is not a string'

File: src/Parser/RegexMatch.php [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+        }
+
+        return $this->values[$key];
+    }
+
+    public function getInt(string $key): int
+    {
+        if (!array_key_exists($key, $this->values) || !is_numeric($this->values[$key])) {
+            throw new RuntimeException(sprintf('key "%s" is not numeric', $key));
```
Comment (guillaume-sainthillier) [inline src/Parser/RegexMatch.php:61]: Same here for `array_key_exists `exception message

**Review by guillaume-sainthillier — COMMENTED** [review]
For my information why using static methods to PHPUNIT::assertXXX ?

File: src/Parser/RegexMatch.php (hunk 1 above)
Comment (fezfez) [reply src/Parser/RegexMatch.php]: done

File: src/Parser/RegexMatch.php (hunk 2 above)
Comment (fezfez) [reply src/Parser/RegexMatch.php:52]: done

File: src/Parser/RegexMatch.php (hunk 3 above)
Comment (fezfez) [reply src/Parser/RegexMatch.php:61]: done

**Comment by fezfez** [comment]
> For my information why using static methods to PHPUNIT::assertXXX ?

Most of method assertXX are static so i use theme as static :)

**Comment by fezfez** [comment]
@guillaume-sainthillier an interesting thing would be to add badge (coverage and type coverage) like https://github.com/azjezz/psl after merging this :)

**Pull request merged**